### PR TITLE
Fix issues

### PR DESCRIPTION
**Pull request by TheCovenant**

### CODE REVIEW

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterComponent/FilterComponent.tsx [hunk 1]
```diff
@@ -192,8 +192,8 @@ const FilterComponentImpl: React.FunctionComponent<{featureType: string}> = ({
             );
             return <React.Fragment key={attributeName} />;
           }
-          //If filters have been set
-          case MapFeatureTypeState.FILTERS_SET: {
+          //If filters have been initialized and all work done for setting up a feature type has completed.
+          case MapFeatureTypeState.FINISHED_SETUP: {
             return returnFilterComponent(
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterComponent/FilterComponent.tsx:197]: Functions that take more than one parameter should use keyword arguments only e.g.,

```
const myFunc = (kwds: {x: string, y: string}) => {
  const {x, y} = kwds;
...
}

myFunc({x: "x", y: "y"})
```

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 2]
```diff
@@ -94,6 +94,28 @@ const MapImpl: React.FunctionComponent = () => {
             transmissionPowerString: feature.transmissionPower
               ? feature.transmissionPower.toString() + " dB"
               : undefined,
+            Y:
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:97]: Or just use my parsedWkt from the server? Can do it another PR.

Comment (gordom6) [reply app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:97]: If you're going to do this in the client at least use a regular expression, this code is hideous. You only need to call the regex once, too.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 3]
```diff
@@ -185,54 +209,88 @@ const MapImpl: React.FunctionComponent = () => {
...
-            for (var x = 0; x < 5; ++x) {
-              /*
+
+            let featureTypeStateOfFeatureType =
+              state.featuresByType[featureType].featureTypeState;
+            switch (featureTypeStateOfFeatureType) {
+              //Check if filters need to be added for this FeatureType
+              case MapFeatureTypeState.NEEDS_FILTERS: {
+                //Dispatch the addFilter action 5 times (1 for each of frequency, timeStamp, transmissionPower, label, and locality)
+                for (var x = 0; x < 5; ++x) {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:227]: This is very fragile. It should be a count somewhere, maybe relative the available attribute strategies?

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 4]
```diff
@@ -185,54 +209,88 @@ const MapImpl: React.FunctionComponent = () => {
...
+                const keplerFieldsOfFeatureType =
+                  keplerState.map.visState.datasets[featureType].fields;
 
-            const interactionConfigCopy =
-              keplerState.map.visState.interactionConfig;
-            interactionConfigCopy.tooltip.config.fieldsToShow[featureType] =
-              featureTypeStrategy.fieldsToShowOnPopup;
-            dispatch(interactionConfigChange(interactionConfigCopy));
+                /*Get the interactionConfig of kepler. This is used to change fieldsToShowOnPopup*/
+                const keplerInteractionConfigCopy =
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:276]: It's not a copy..?

Comment (gordom6) [reply app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:276]: But it should be?

File: app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/FeatureTypeStrategy.ts [hunk 5]
```diff
@@ -1,10 +1,22 @@
...
+import {FeaturesByType} from "../states/map/FeaturesByType";
 
 export interface FeatureTypeStrategy {
   readonly name: FeatureType;
   readonly childFeatureTypes: FeatureType[];
   readonly expandable: boolean;
   readonly fieldsToShowOnPopup: FeatureAttributeName[];
   getClickedQuery(clickedFeatureUri: string): FeatureQuery | null;
+  dispatchLayerConfigurationActions(
+    keplerLayerOfFeatureType: any,
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/FeatureTypeStrategy.ts:13]: kwds: { ... }

Comment (gordom6) [reply app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/FeatureTypeStrategy.ts:13]: The other nice thing about kwds is that you can put them in an interface and reuse them.
kwds: LayerConfigurationActionParameters
interface LayerConfigurationActionParameters { ... }
This is called a "parameter object": https://wiki.c2.com/?ParameterObject

File: app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/PointFeatureTypeStrategy.ts [hunk 6]
```diff
@@ -1,11 +1,23 @@
...
 import {FeatureAttributeName} from "../../states/map/FeatureAttributeName";
+import {Dispatch} from "redux";
+import {FeaturesByType} from "../../states/map/FeaturesByType";
 
 export abstract class PointFeatureTypeStrategy implements FeatureTypeStrategy {
   getClickedQuery(clickedFeatureUri: string): null {
     return null;
   }
+  abstract dispatchLayerConfigurationActions(
+    keplerLayerOfFeatureType: any,
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/PointFeatureTypeStrategy.ts:12]: Reuse with the FeatureTypeStrategy

File: app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/PolicyFeatureTypeStrategy.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import {PointFeatureTypeStrategy} from "./PointFeatureTypeStrategy";
+import {FeatureType} from "../../api/graphqlGlobalTypes";
+import {FeatureAttributeName} from "../../states/map/FeatureAttributeName";
+import {FeaturesByType} from "../../states/map/FeaturesByType";
+import {Dispatch} from "redux";
+
+export class PolicyFeatureTypeStrategy extends PointFeatureTypeStrategy {
+  readonly name = FeatureType.Policy;
+  dispatchLayerConfigurationActions(
+    keplerLayerOfFeatureType: any,
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/PolicyFeatureTypeStrategy.ts:10]: Reuse the interface again here.

File: app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PolygonFeatureTypeStrategy/PolygonFeatureTypeStrategy.ts [hunk 8]
```diff
@@ -1,9 +1,41 @@
...
 import {FeatureAttributeName} from "../../states/map/FeatureAttributeName";
+import {Dispatch} from "redux";
+import {FeaturesByType} from "../../states/map/FeaturesByType";
+import {MapFeatureTypeState} from "../../states/map/MapFeatureTypeState";
+import {layerConfigChange, interactionConfigChange} from "kepler.gl/actions";
 
 export abstract class PolygonFeatureTypeStrategy
   implements FeatureTypeStrategy {
+  dispatchLayerConfigurationActions(
+    keplerLayerOfFeatureType: any,
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PolygonFeatureTypeStrategy/PolygonFeatureTypeStrategy.ts:12]: Reuse reuse reuse.

File: app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PolygonFeatureTypeStrategy/PolygonFeatureTypeStrategy.ts [hunk 9]
```diff
@@ -1,9 +1,41 @@
...
+    keplerFiltersOfFeatureType: any,
+    keplerFieldsOfFeatureType: any,
+    keplerInteractionConfigOfFeatureType: any,
+    dispatch: Dispatch<any>,
+    featuresByType: {
+      [featureType: string]: FeaturesByType;
+    }
+  ): void {
+    switch (featuresByType[this.name].featureTypeState) {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PolygonFeatureTypeStrategy/PolygonFeatureTypeStrategy.ts:21]: The implementations all appear to be doing this expression, just pass it in instead of featuresByType?

File: app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts [hunk 10]
```diff
@@ -243,14 +244,12 @@ export const mapReducer = (state: MapState, action: BaseAction): MapState => {
         setAllFilterIndexesNull(attributeStatesOfFeatureType);
 
         if (
-          //If the filters of the attributes of a FeatureType have been added/set
-          featuresByTypeOfType.featureTypeState ===
-            MapFeatureTypeState.FILTERS_ADDED ||
-          featuresByTypeOfType.featureTypeState ===
-            MapFeatureTypeState.FILTERS_SET
+          //If features of that type are present on the map
+          featuresByTypeOfType.featureTypeState !=
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts:248]: !== ?

File: app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts [hunk 11]
```diff
@@ -395,8 +394,77 @@ export const mapReducer = (state: MapState, action: BaseAction): MapState => {
...
+            MapFeatureTypeState.NEEDS_LAYER_CHANGE;
+        } else {
+          featuresByTypeOfFeatureType.featureTypeState =
+            MapFeatureTypeState.FINISHED_SETUP;
+        }
+      }
+      /*Checking to see if the action was used to change the 'columns' of the layer. If so, then it signifies 
+      that we are in MapFeatureTypeState.NEEDS_LNG_AND_LAT and it is time to move to NEEDS_3D_ENABLED*/
+      if (Object.keys(layerConfigChangeAction.newConfig).includes("columns")) {
+        result.featuresByType[FeatureType.Transmission].featureTypeState =
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducers/map/mapReducer.ts:430]: Hard-coding a FeatureType here smells. This should be abstracted into the strategy.

File: app/geo/gui/src/ts/twxplore/gui/geo/states/map/MapFeature.ts [hunk 12]
```diff
@@ -3,6 +3,8 @@ import {MapFeaturesQuery_features} from "../../api/queries/types/MapFeaturesQuer
 
 export interface MapFeature extends MapFeaturesQuery_features {
   state: MapFeatureState;
+  Y?: number;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/states/map/MapFeature.ts:6]: Proper casing, please.

File: app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/TransmitterFeatureTypeStrategy.ts [hunk 13]
```diff
@@ -1,9 +1,41 @@
...
 import {FeatureType} from "../../api/graphqlGlobalTypes";
 import {FeatureAttributeName} from "../../states/map/FeatureAttributeName";
+import {FeaturesByType} from "../../states/map/FeaturesByType";
+import {Dispatch} from "redux";
+import {MapFeatureTypeState} from "../../states/map/MapFeatureTypeState";
+import {layerConfigChange, interactionConfigChange} from "kepler.gl/actions";
 
 export class TransmitterFeatureTypeStrategy extends PointFeatureTypeStrategy {
   readonly name = FeatureType.Transmitter;
+  dispatchLayerConfigurationActions(
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/TransmitterFeatureTypeStrategy.ts:11]: Reuse again.

File: app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/PointFeatureTypeStrategy.ts [hunk 14]
```diff
@@ -3,23 +3,26 @@ import {FeatureType} from "../../api/graphqlGlobalTypes";
...
   getClickedQuery(clickedFeatureUri: string): null {
     return null;
   }
-  abstract dispatchLayerConfigurationActions(
-    keplerLayerOfFeatureType: any,
-    keplerFiltersOfFeatureType: any,
-    keplerFieldsOfFeatureType: any,
-    keplerInteractionConfigOfFeatureType: any,
-    dispatch: Dispatch<any>,
+  abstract dispatchLayerConfigurationActions(kwds: {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/PointFeatureTypeStrategy.ts:12]: Not sure what's going on here. You declared the parameter object interface but aren't using it? If there are optional members then add them as ? to the interface, or have another interface that extends the first one.

File: app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/TransmissionFeatureTypeStrategy.ts [hunk 15]
```diff
@@ -12,45 +10,50 @@ import {
...
       case MapFeatureTypeState.NEEDS_LNG_AND_LAT: {
         const latFieldIdx = keplerFieldsOfFeatureType.findIndex(
-          (keplerField: {id: string}) => keplerField.id === "Y"
+          (keplerField: {id: string}) =>
+            keplerField.id === FeatureAttributeName.y
         );
         const lngFieldIdx = keplerFieldsOfFeatureType.findIndex(
-          (keplerField: {id: string}) => keplerField.id === "X"
+          (keplerField: {id: string}) =>
+            keplerField.id === FeatureAttributeName.x
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/TransmissionFeatureTypeStrategy.ts:51]: I like this a lot better than the magic string.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 16]
```diff
@@ -78,44 +82,29 @@ const MapImpl: React.FunctionComponent = () => {
...
-                ? parseFloat(
-                    feature.geometry.wkt
-                      .replace("POINT", "")
-                      .replace("(", "")
-                      .trim()
-                      .replace(")", "")
-                      .split(" ")[0]
-                  )
-                : undefined,
+            x: (feature.geometry
```
[the remark this answers is not in the document]
Comment (TheCovenant) [reply app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:100]: Not all features are points so if a feature isn't a point, then x and y will be undefined. Which is fine behavior for me because x and y should only be defined for features that are points. I only use x and y for features that are points.

File: app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/PointFeatureTypeStrategy.ts (hunk 14 above)
Comment (TheCovenant) [reply app/geo/gui/src/ts/twxplore/gui/geo/featureTypeStrategies/PointFeatureTypeStrategy/PointFeatureTypeStrategy.ts:12]: Oh big mistake. I created the interface after and I forgot to replace the parameter object

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 17]
```diff
@@ -273,20 +298,20 @@ const MapImpl: React.FunctionComponent = () => {
                   keplerState.map.visState.datasets[featureType].fields;
 
                 /*Get the interactionConfig of kepler. This is used to change fieldsToShowOnPopup*/
-                const keplerInteractionConfigCopy =
+                const keplerInteractionConfig =
```
[the remark this answers is not in the document]
Comment (TheCovenant) [reply app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:299]: Ah yes. I thought useSelector created deep copies.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 18]
```diff
@@ -298,17 +301,20 @@ const MapImpl: React.FunctionComponent = () => {
                   keplerState.map.visState.datasets[featureType].fields;
 
                 /*Get the interactionConfig of kepler. This is used to change fieldsToShowOnPopup*/
-                const keplerInteractionConfig =
-                  keplerState.map.visState.interactionConfig;
-                keplerInteractionConfig.tooltip.config.fieldsToShow[
+                const keplerInteractionConfigCopy = Object.assign(
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:304]: This is a shallow copy, but I assumed that's OK.
